cocotb/types/array.py

```diff
@@ -0,0 +1,314 @@
+
+    def __eq__(self, other: Any) -> bool:
+        try:
+            return all(a == b for a, b in zip_longest(self, other, fillvalue=object()))
+        except TypeError:
+            return NotImplemented
+
+    @overload
+    def __getitem__(self, item: int) -> Any:
+        pass  # pragma: no cover
+
+    @overload
+    def __getitem__(self, item: slice) -> "Array":
+        pass  # pragma: no cover
+
+    def __getitem__(self, item):
+        if isinstance(item, int):
+            idx = self._translate_index(item)
+            return self._value[idx]
+        elif isinstance(item, slice):
+            start = item.start or self.left
+            stop = item.stop or self.right
+            if item.step is not None:
+                raise IndexError("do not specify step")
+            start_i = self._translate_index(start)
+            stop_i = self._translate_index(stop)
+            if start_i > stop_i:
+                raise IndexError("slice direction does not match array direction")
+            value = self._value[start_i : stop_i + 1]
+            range = Range(start, self.direction, stop)
+            return type(self)(value=value, range=range)
+        raise TypeError(
+            "indexes must be ints or slices, not {}".format(type(item).__name__)
+        )
+
+    @overload
+    def __setitem__(self, item: int, value: Any) -> None:
+        pass  # pragma: no cover
+
+    @overload
+    def __setitem__(self, item: slice, value: Iterable[Any]) -> None:
+        pass  # pragma: no cover
+
+    def __setitem__(self, item, value):
+        if isinstance(item, int):
+            idx = self._translate_index(item)
+            value = self._construct_element(value)
+            self._value[idx] = value
+        elif isinstance(item, slice):
+            start = item.start or self.left
+            stop = item.stop or self.right
+            if item.step is not None:
+                raise IndexError("do not specify step")
+            start_i = self._translate_index(start)
+            stop_i = self._translate_index(stop)
+            if start_i > stop_i:
+                raise IndexError("slice direction does not match array direction")
+            value = self._construct_value(value)
+            if len(value) != (stop_i - start_i + 1):
+                raise ValueError(
+                    "value  of length {!r} will not fit in slice [{}:{}]".format(
+                        len(value), start, stop
+                    )
+                )
+            self._value[start_i : stop_i + 1] = value
+        else:
+            raise TypeError(
+                "indexes must be ints or slices, not {}".format(type(item).__name__)
+            )
+
+    def __repr__(self) -> str:
+        return "{}({!r}, {!r})".format(type(self).__name__, self._value, self._range)
+
+    def index(
+        self, value: Any, start: Optional[int] = None, stop: Optional[int] = None
+    ) -> int:
+        """Return index of first occurrence of *value*."""
```

Comment:
```suggestion
        """Return index of first occurrence of *value*.
        
        Search only within *start* and *stop* if given.
        """
```
(or something like that.)